src/lib.rs

```diff
@@ -1198,36 +1174,68 @@ decl_module! {
         }
 
         /// submit a poll
-        // fn submit_poll(origin, thread_id: T::ThreadId, poll_value: PollData) -> dispatch::Result {
-        //     // Check that its a valid signature
-        //     let who = ensure_signed(origin)?;
-
-        //     // get forum user id.
-        //     let forum_user_id = Self::ensure_is_forum_member_with_correct_account(&who)?;
-
-        //     // Get thread
-        //     let thread = Self::ensure_thread_exists(&thread_id)?;
-
-        //     // Make sure poll exist and not expired
-        //     match thread.poll {
-        //         None => Err(ERROR_POLL_NOT_EXIST),
-        //         Some(poll) => {
-        //             if poll.end_time < <timestamp::Module<T>>::now() {
-        //                 Err(ERROR_POLL_COMMIT_EXPIRED)
-        //             } else {
-        //                 Self::ensure_poll_data_valid(&poll, poll_value)?;
-        //                 <PollDataByAccount<T>>::insert(thread_id, forum_user_id, poll_value);
-        //                 if <PollStatistics<T>>::exists(thread_id, poll_value) {
-        //                     <PollStatistics<T>>::insert(thread_id, poll_value, 1);
-        //                 } else {
-        //                     let old_value = <PollStatistics<T>>::get(thread_id, poll_value);
-        //                     <PollStatistics<T>>::insert(thread_id, poll_value, old_value + 1);
-        //                 }
-        //                 Ok(())
-        //             }
-        //         }
-        //     }
-        // }
+        fn submit_poll(origin, forum_user_id: T::ForumUserId, thread_id: T::ThreadId, index: u32) -> dispatch::Result {
+            // Check that its a valid signature
+            let who = ensure_signed(origin)?;
+
+            // get forum user id.
+            Self::ensure_is_forum_member_with_correct_account(&who, &forum_user_id)?;
+
+            // Get thread
+            let thread = Self::ensure_thread_exists(&thread_id)?;
+
+            // Make sure poll exist
+            match thread.poll {
+                None => Err(ERROR_POLL_NOT_EXIST),
+                Some(poll) => {
+                    // Poll not expired
+                    if poll.end_time < <timestamp::Module<T>>::now() {
+                        Err(ERROR_POLL_COMMIT_EXPIRED)
+                    } else {
+                        let alternative_length = poll.poll_alternatives.len();
+                        // The selected alternative index is valid
+                        if index as usize >= alternative_length {
+                            Err(ERROR_POLL_DATA)
+                        } else {
+                            // Update thread data
+                            <ThreadById<T>>::mutate(thread_id, |value| {
+                                // Store new poll alternative statistics
+                                let mut new_vote_count = poll.poll_alternatives.clone();
```

Comment:
This variable has an incorrect name, probably coming from its mutable nature where the semantics are changing underway.

1. Avoid doing as much computation or calculation as possible inside of mutation call, try to do outside in, and only do write in mutation, if possible.

2. This can be done without mutable variable like this

Reply:
1. accepted, I will move out of mutate call.
2. accepted, just forgot functional programming sometime.